crawler/crawler.go

```diff
@@ -0,0 +1,154 @@
+package crawler
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	logging "github.com/ipfs/go-log"
+
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+)
+
+var logger = logging.Logger("dht-crawler")
+
+type Crawler struct {
+	parallelism int
+	host        host.Host
+	dhtRPC      *dht.ProtocolMessenger
+}
+
+func New(host host.Host, opts ...Option) (*Crawler, error) {
+	o := new(options)
+	if err := defaults(o); err != nil {
+		return nil, err
+	}
+	for _, opt := range opts {
+		if err := opt(o); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Crawler{
+		parallelism: o.parallelism,
+		host:        host,
+		dhtRPC:      dht.NewProtocolMessenger(host, o.protocols, nil),
+	}, nil
+}
+
+type HandleQueryResult func(p peer.ID, rtPeers []*peer.AddrInfo)
+type HandleQueryFail func(p peer.ID, err error)
+
+func (c *Crawler) Run(ctx context.Context, startingPeers []*peer.AddrInfo, handleSuccess HandleQueryResult, handleFail HandleQueryFail) {
+	jobs := make(chan peer.ID, 1)
+	results := make(chan *queryResult, 1)
+
+	// Start worker goroutines
+	var wg sync.WaitGroup
+	wg.Add(c.parallelism)
+	for i := 0; i < c.parallelism; i++ {
+		go func() {
+			for p := range jobs {
+				res := queryPeer(ctx, c.host, c.dhtRPC, p)
+				results <- res
+			}
+		}()
+	}
+
+	defer wg.Done()
```

Comment:
the go routines above don't decrement wg when they finish, so this won't work as expected